tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -48,84 +48,97 @@ astDeclarationsForLanguage language filePath = do
-  where child = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT (toName named n))
+constructorForSumChoice :: String -> Name -> TreeSitter.Deserialize.Type -> Q Con
+constructorForSumChoice str typeParameterName (MkType (DatatypeName n) named) = normalC (toName named (n ++ str)) [child]
+  where child = TH.bangType strictness (appT (conT (toName named n)) (varT typeParameterName))
 
 -- | Build Q Constructor for product types (nodes with fields)
-ctorForProductType :: String -> Maybe Children -> [(String, Field)] -> Q Con
-ctorForProductType constructorName children fields = recC (toName Named constructorName) lists where
-  lists = fieldList ++ childList
-  fieldList = fmap (uncurry toVarBangType) fields
-  childList = toList $ fmap toVarBangTypeChild children
-  toVarBangType name (MkField required fieldTypes mult) =
-    let fieldName = mkName . addTickIfNecessary . removeUnderscore $ name
-        strictness = TH.bang noSourceUnpackedness noSourceStrictness
-        ftypes = fieldTypesToNestedEither fieldTypes
-        fieldContents = case (required, mult) of
-          (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
-          (Required, Single) -> ftypes
-          (Optional, Multiple) -> appT (conT ''[]) ftypes
-          (Optional, Single) -> appT (conT ''Maybe) ftypes
-    in TH.varBangType fieldName (TH.bangType strictness fieldContents)
-  toVarBangTypeChild (MkChildren field) = toVarBangType "extra_children" field
+ctorForProductType :: String -> Name -> Maybe Children -> [(String, Field)] -> Q Con
+ctorForProductType constructorName typeParameterName children fields = ctorForTypes Named constructorName lists where
+  lists = annotation : fieldList ++ childList
+  annotation = ("ann", varT typeParameterName)
+  fieldList = map (fmap toType) fields
+  childList = toList $ fmap toTypeChild children
+  toType (MkField required fieldTypes mult) =
+    let ftypes = fieldTypesToNestedEither fieldTypes typeParameterName
+    in case (required, mult) of
+      (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
+      (Required, Single) -> ftypes
+      (Optional, Multiple) -> appT (conT ''[]) ftypes
+      (Optional, Single) -> appT (conT ''Maybe) ftypes
+  toTypeChild (MkChildren field) = ("extra_children", toType field)
 
 
 -- | Build Q Constructor for leaf types (nodes with no fields or subtypes)
-ctorForLeafType :: Named -> DatatypeName -> Q Con
-ctorForLeafType Anonymous (DatatypeName name) = normalC (toName Anonymous name) []
-ctorForLeafType Named (DatatypeName name) = recC (toName Named name) [leafBytes] where
-  leafBytes = TH.varBangType (mkName "bytes") textValue
-  textValue = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT ''Text)
+ctorForLeafType :: Named -> DatatypeName -> Name -> Q Con
+ctorForLeafType named datatypeName typeParameterName =
+  case (named, datatypeName) of
+    (Anonymous, DatatypeName name) -> ctorForTypes Anonymous name [annotation, ("bytes", conT ''Text)]
```

Comment:
Okay, it looks like this has indeed added `bytes` fields to anonymous datatypes, which is unnecessary and even counterproductive—the node representing a token like `+` is only ever differentiated by its annotation, never its bytes (i.e. `+` is always just `+`, wherever it occurs).

Reply:
Ah, I see—I led @aymannadeem slightly wrong, then. It looks like anonymous nodes should be newtypes over annotation values.

Reply:
> the node representing a token like + is only ever differentiated by its annotation, never its bytes (i.e. + is always just +, wherever it occurs).

Is this always true? Could there be multiple rules that mapped to a single token?

Reply:
No; tokens are constant.

Reply:
> Could there be multiple recognized byte sequences that mapped to the same token?

No; tokens are constant. Any other sequence of byte sequence would necessarily map to a different token.

Reply:
(Tho note that they’re not the only constant nodes, e.g. cf #136 describing named rules wrapping tokens.)

Reply:
@patrickt not your fault at all, I proposed it because I misunderstood the end goal. I double-checked reverting it back to a datatype with you before @robrix got in and told me to stop with the `bytes`.